File: src/cppinterface/cb_summodel.jl
@doc raw"""
    cb_clear!(self::CBSumModel)

resets the SumModel to its initial state, in particular it removes but does not delete any Models added in add_model() (their AFTs are deleted if not explicitly denied)
"""
cb_clear!(self::CBSumModel) = @ccall libcb.cb_summodel_clear(self.data::Ptr{Cvoid})::Cvoid

@doc raw"""
    CBSumModel()

constructor
"""
CBSumModel() = CBSumModel(@ccall libcb.cb_summodel_new()::Ptr{Cvoid})

@doc raw"""
    cb_model(self::CBSumModel, fo::Union{<:CBFunctionObject,Nothing})

returns the submodel for FunctionObject fo if it in this model, otherwise 0
"""
cb_model(self::CBSumModel, fo::Union{<:CBFunctionObject,Nothing}) = CBSumBlockModel(@ccall libcb.cb_summodel_model(self.data::Ptr{Cvoid}, (isnothing(fo) ? C_NULL : fo.data)::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_nsubmodels(self::CBSumModel)

returns the number of submodels in this model
"""
cb_nsubmodels(self::CBSumModel) = @ccall libcb.cb_summodel_nsubmodels(self.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_add_model!(self::CBSumModel, model::Union{<:CBSumBlockModel,Nothing})

adds the @a model as submodel to this model (if this model may have submodels); if any error occurs the model is not added
"""
cb_add_model!(self::CBSumModel, model::Union{<:CBSumBlockModel,Nothing}) = @ccall libcb.cb_summodel_add_model(self.data::Ptr{Cvoid}, (isnothing(model) ? C_NULL : model.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_remove_model!(self::CBSumModel, fo::Union{<:CBFunctionObject,Nothing})

remove the submodel identified by @a fo from this model, this does NOT destruct the model. It returns the pointer to the model if there is one, otherwise 0
"""
cb_remove_model!(self::CBSumModel, fo::Union{<:CBFunctionObject,Nothing}) = CBSumBlockModel(@ccall libcb.cb_summodel_remove_model(self.data::Ptr{Cvoid}, (isnothing(fo) ? C_NULL : fo.data)::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_eval_function!(self::CBSumModel, y_id::Integer, y::CBMatrix, nullstep_bound::Real, relprec::Real)

see BundleModel::eval_function
"""
function cb_eval_function!(self::CBSumModel, y_id::Integer, y::CBMatrix, nullstep_bound::Real, relprec::Real)
    ub = Ref{Float64}()
    ub_fid = Ref{Int}()
    @ccall libcb.cb_summodel_eval_function(self.data::Ptr{Cvoid}, ub_fid::Ref{Int}, ub::Ref{Float64}, y_id::Cint, y.data::Ptr{Cvoid}, nullstep_bound::Cdouble, relprec::Cdouble)::Cint
    return ub_fid[], ub[]
end

@doc raw"""
    cb_eval_model!(self::CBSumModel, y_id::Integer, y::CBMatrix, relprec::Real)

see BundleModel::eval_model
"""
function cb_eval_model!(self::CBSumModel, y_id::Integer, y::CBMatrix, relprec::Real)
    lb = Ref{Float64}()
    @ccall libcb.cb_summodel_eval_model(self.data::Ptr{Cvoid}, lb::Ref{Float64}, y_id::Cint, y.data::Ptr{Cvoid}, relprec::Cdouble)::Cint
    return lb[]
end

@doc raw"""
    cb_update_model!(self::CBSumModel, model_update::CBModelUpdate, center_id::Integer, center_y::CBMatrix, cand_id::Integer, cand_y::CBMatrix, model_maxviol::Real, H::CBBundleProxObject)

see BundleModel::update_model
"""
cb_update_model!(self::CBSumModel, model_update::CBModelUpdate, center_id::Integer, center_y::CBMatrix, cand_id::Integer, cand_y::CBMatrix, model_maxviol::Real, H::CBBundleProxObject) = @ccall libcb.cb_summodel_update_model(self.data::Ptr{Cvoid}, model_update::Cint, center_id::Cint, center_y.data::Ptr{Cvoid}, cand_id::Cint, cand_y.data::Ptr{Cvoid}, model_maxviol::Cdouble, H.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_synchronize_ids!(self::CBSumModel, new_center_id::Integer, old_center_id::Integer, new_cand_id::Integer, old_cand_id::Integer)

see BundleModel::synchronize_ids
"""
function cb_synchronize_ids!(self::CBSumModel, new_center_id::Integer, old_center_id::Integer, new_cand_id::Integer, old_cand_id::Integer)
    new_aggregate_id = Ref{Int}()
    new_cand_ub_fid = Ref{Int}()
    new_center_ub_fid = Ref{Int}()
    @ccall libcb.cb_summodel_synchronize_ids(self.data::Ptr{Cvoid}, new_center_ub_fid::Ref{Int}, new_center_id::Cint, old_center_id::Cint, new_cand_ub_fid::Ref{Int}, new_cand_id::Cint, old_cand_id::Cint, new_aggregate_id::Ref{Int})::Cint
    return new_center_ub_fid[], new_cand_ub_fid[], new_aggregate_id[]
end

@doc raw"""
    cb_center_modified!(self::CBSumModel, center_id::Integer)

see BundleModel::center_modified
"""
function cb_center_modified!(self::CBSumModel, center_id::Integer)
    function_modification_id = Ref{Int}()
    Bool(@ccall libcb.cb_summodel_center_modified(self.data::Ptr{Cvoid}, function_modification_id::Ref{Int}, center_id::Cint)::Cint)
    return function_modification_id[]
end

@doc raw"""
    cb_recompute_center!(self::CBSumModel, center_id::Integer, y::CBMatrix, accept_only_higher_values::Bool = false, relprec::Real = -1.)

see BundleModel::recompute_center
"""
function cb_recompute_center!(self::CBSumModel, center_id::Integer, y::CBMatrix, accept_only_higher_values::Bool = false, relprec::Real = -1.)
    new_center_ub = Ref{Float64}()
    new_center_ub_fid = Ref{Int}()
    @ccall libcb.cb_summodel_recompute_center(self.data::Ptr{Cvoid}, new_center_ub_fid::Ref{Int}, new_center_ub::Ref{Float64}, center_id::Cint, y.data::Ptr{Cvoid}, accept_only_higher_values::Cint, relprec::Cdouble)::Cint
    return new_center_ub_fid[], new_center_ub[]
end

@doc raw"""
    cb_model_aggregate_modified!(self::CBSumModel, old_model_aggregate_id::Integer)

see BundleModel::model_aggregate_modified
"""
cb_model_aggregate_modified!(self::CBSumModel, old_model_aggregate_id::Integer) = Bool(@ccall libcb.cb_summodel_model_aggregate_modified(self.data::Ptr{Cvoid}, old_model_aggregate_id::Cint)::Cint)

@doc raw"""
    cb_provide_model_aggregate!(self::CBSumModel, y_id::Integer, y::CBMatrix)

see BundleModel::provide_model_aggregate
"""
cb_provide_model_aggregate!(self::CBSumModel, y_id::Integer, y::CBMatrix) = @ccall libcb.cb_summodel_provide_model_aggregate(self.data::Ptr{Cvoid}, y_id::Cint, y.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_add_variable_metric!(self::CBSumModel, H::CBVariableMetric, y_id::Integer, y::CBMatrix, descent_step::Bool, weightu::Real, model_maxviol::Real, indices::Union{<:CBIndexmatrix,Nothing} = nothing)

see BundleModel::add_dynamic_scaling
"""
cb_add_variable_metric!(self::CBSumModel, H::CBVariableMetric, y_id::Integer, y::CBMatrix, descent_step::Bool, weightu::Real, model_maxviol::Real, indices::Union{<:CBIndexmatrix,Nothing} = nothing) = @ccall libcb.cb_summodel_add_variable_metric(self.data::Ptr{Cvoid}, H.data::Ptr{Cvoid}, y_id::Cint, y.data::Ptr{Cvoid}, descent_step::Cint, weightu::Cdouble, model_maxviol::Cdouble, (isnothing(indices) ? C_NULL : indices.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_check_center_validity_by_candidate!(self::CBSumModel, cand_minorant_is_below::Bool, center_id::Integer, center_y::CBMatrix)

see BundleModel::check_center_validity_by_candidate
"""
cb_check_center_validity_by_candidate!(self::CBSumModel, cand_minorant_is_below::Bool, center_id::Integer, center_y::CBMatrix) = @ccall libcb.cb_summodel_check_center_validity_by_candidate(self.data::Ptr{Cvoid}, cand_minorant_is_below::Ref{Cint}, center_id::Cint, center_y.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_oracle_object!(self::CBSumModel)

as AFTModel has no oracle of its own, this returns the dummy oracle
"""
cb_get_oracle_object!(self::CBSumModel) = CBModifiableOracleObject(@ccall libcb.cb_summodel_get_oracle_object(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_make_model_aggregate!(self::CBSumModel, penalty_parameter_increased::Bool, keep_penalty_fixed::Bool)

see SumBlockModel::make_model_aggregate
"""
cb_make_model_aggregate!(self::CBSumModel, penalty_parameter_increased::Bool, keep_penalty_fixed::Bool) = @ccall libcb.cb_summodel_make_model_aggregate(self.data::Ptr{Cvoid}, penalty_parameter_increased::Ref{Cint}, keep_penalty_fixed::Cint)::Cint

@doc raw"""
    cb_lb_function!(self::CBSumModel, y_id::Integer, y::CBMatrix)

see SumBlockModel::lb_function()
"""
cb_lb_function!(self::CBSumModel, y_id::Integer, y::CBMatrix) = @ccall libcb.cb_summodel_lb_function(self.data::Ptr{Cvoid}, y_id::Cint, y.data::Ptr{Cvoid})::Cdouble

@doc raw"""
    cb_get_function_minorant!(self::CBSumModel, minorant::CBMinorantPointer, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing)

see SumBlockModel::get_function_minorant()
"""
cb_get_function_minorant!(self::CBSumModel, minorant::CBMinorantPointer, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing) = @ccall libcb.cb_summodel_get_function_minorant(self.data::Ptr{Cvoid}, minorant.data::Ptr{Cvoid}, (isnothing(aft) ? C_NULL : aft.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_center_minorant!(self::CBSumModel, minorant::CBMinorantPointer, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing)

see SumBlockModel::get_center_minorant()
"""
cb_get_center_minorant!(self::CBSumModel, minorant::CBMinorantPointer, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing) = @ccall libcb.cb_summodel_get_center_minorant(self.data::Ptr{Cvoid}, minorant.data::Ptr{Cvoid}, (isnothing(aft) ? C_NULL : aft.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_adjust_multiplier!(self::CBSumModel, values_may_have_changed::Bool)

see SumBlockModel::adjust_multiplier()
"""
cb_adjust_multiplier!(self::CBSumModel, values_may_have_changed::Bool) = @ccall libcb.cb_summodel_adjust_multiplier(self.data::Ptr{Cvoid}, values_may_have_changed::Ref{Cint})::Cint

@doc raw"""
    cb_sumbundle_mode!(self::CBSumModel, bh::Union{<:CBSumBundleHandler,Nothing} = nothing, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing)

see SumBlockModel::sumbundle_mode
"""
function cb_sumbundle_mode!(self::CBSumModel, bh::Union{<:CBSumBundleHandler,Nothing} = nothing, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing)
    mode = Ref{CBMode}()
    @ccall libcb.cb_summodel_sumbundle_mode(self.data::Ptr{Cvoid}, mode::Ref{CBMode}, (isnothing(bh) ? C_NULL : bh.data)::Ptr{Cvoid}, (isnothing(aft) ? C_NULL : aft.data)::Ptr{Cvoid})::Cint
    return mode[]
end

@doc raw"""
    cb_start_sumaugmodel!(self::CBSumModel, blockp::CBQPModelDataPointer, cand_id::Integer, cand_y::CBMatrix, indices::Union{<:CBIndexmatrix,Nothing} = nothing, bh::Union{<:CBSumBundleHandler,Nothing} = nothing, mode::CBMode = cbm_inactive, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing)

see BundleModel::start_augmodel() for the first four parameters; for the others see sumbundle_contribution()
"""
cb_start_sumaugmodel!(self::CBSumModel, blockp::CBQPModelDataPointer, cand_id::Integer, cand_y::CBMatrix, indices::Union{<:CBIndexmatrix,Nothing} = nothing, bh::Union{<:CBSumBundleHandler,Nothing} = nothing, mode::CBMode = cbm_inactive, aft::Union{<:CBAffineFunctionTransformation,Nothing} = nothing) = @ccall libcb.cb_summodel_start_sumaugmodel(self.data::Ptr{Cvoid}, blockp.data::Ptr{Cvoid}, cand_id::Cint, cand_y.data::Ptr{Cvoid}, (isnothing(indices) ? C_NULL : indices.data)::Ptr{Cvoid}, (isnothing(bh) ? C_NULL : bh.data)::Ptr{Cvoid}, mode::Cint, (isnothing(aft) ? C_NULL : aft.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_data!(self::CBSumModel)

see SumBlockModel::get_data()
"""
cb_get_data!(self::CBSumModel) = CBBundleData(@ccall libcb.cb_summodel_get_data(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_data(self::CBSumModel)

see SumBlockModel::get_data()
"""
cb_get_data(self::CBSumModel) = CBBundleData(@ccall libcb.cb_summodel_get_data2(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_set_data!(self::CBSumModel, bd::Union{<:CBBundleData,Nothing})

see SumBlockModel::set_data()
"""
cb_set_data!(self::CBSumModel, bd::Union{<:CBBundleData,Nothing}) = @ccall libcb.cb_summodel_set_data(self.data::Ptr{Cvoid}, (isnothing(bd) ? C_NULL : bd.data)::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_approximate_primal(self::CBSumModel)

this has no primals, so it returns 1, see SumBlockModel::get_approximate_primal
"""
cb_get_approximate_primal(self::CBSumModel) = CBPrimalData(@ccall libcb.cb_summodel_get_approximate_primal(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_center_primal(self::CBSumModel)

this has no primals, so it returns 1, see SumBlockModel::get_center_primal
"""
cb_get_center_primal(self::CBSumModel) = CBPrimalData(@ccall libcb.cb_summodel_get_center_primal(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_candidate_primal(self::CBSumModel)

this has no primals, so it returns 1, see SumBlockModel::get_candidate_primal
"""
cb_get_candidate_primal(self::CBSumModel) = CBPrimalData(@ccall libcb.cb_summodel_get_candidate_primal(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_call_primal_extender!(self::CBSumModel, param0::CBPrimalExtender)

this has no primals, so it returns 1, see SumBlockModel::call_primal_extender
"""
cb_call_primal_extender!(self::CBSumModel, param0::CBPrimalExtender) = @ccall libcb.cb_summodel_call_primal_extender(self.data::Ptr{Cvoid}, param0.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_set_bundle_parameters!(self::CBSumModel, bp::CBBundleParameters)

if bp is a SumBundleParametersObject, it is cloned, otherwise the paraemters are copied
"""
cb_set_bundle_parameters!(self::CBSumModel, bp::CBBundleParameters) = @ccall libcb.cb_summodel_set_bundle_parameters(self.data::Ptr{Cvoid}, bp.data::Ptr{Cvoid})::Cint

@doc raw"""
    cb_get_bundle_parameters(self::CBSumModel)

see SumBlockModel::get_bundle_parameters()
"""
cb_get_bundle_parameters(self::CBSumModel) = CBBundleParameters(@ccall libcb.cb_summodel_get_bundle_parameters(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_clear_model!(self::CBSumModel, discard_minorants_only::Bool = false)

see SumBlockModel::clear_model
"""
cb_clear_model!(self::CBSumModel, discard_minorants_only::Bool = false) = @ccall libcb.cb_summodel_clear_model(self.data::Ptr{Cvoid}, discard_minorants_only::Cint)::Cvoid

@doc raw"""
    cb_clear_aggregates!(self::CBSumModel)

see SumBlockModel::clear_aggregates
"""
cb_clear_aggregates!(self::CBSumModel) = @ccall libcb.cb_summodel_clear_aggregates(self.data::Ptr{Cvoid})::Cvoid

@doc raw"""
    cb_set_out!(self::CBSumModel, pril::Integer = 1)

set output and outputlevel of warnings and errors recursively, see CBout
"""
cb_set_out!(self::CBSumModel, pril::Integer = 1) = @ccall libcb.cb_summodel_set_out(self.data::Ptr{Cvoid}, pril::Cint)::Cvoid

@doc raw"""
    cb_get_preeval_time(self::CBSumModel)

see SumBlockModel::get_preeval_time()
"""
cb_get_preeval_time(self::CBSumModel) = CBCH_Tools::Microseconds(@ccall libcb.cb_summodel_new_get_preeval_time(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_eval_time(self::CBSumModel)

see SumBlockModel::get_eval_time()
"""
cb_get_eval_time(self::CBSumModel) = CBCH_Tools::Microseconds(@ccall libcb.cb_summodel_new_get_eval_time(self.data::Ptr{Cvoid})::Ptr{Cvoid})

@doc raw"""
    cb_get_posteval_time(self::CBSumModel)

see SumBlockModel::get_posteval_time()
"""
cb_get_posteval_time(self::CBSumModel) = CBCH_Tools::Microseconds(@ccall libcb.cb_summodel_new_get_posteval_time(self.data::Ptr{Cvoid})::Ptr{Cvoid})

File: ConicBundle/cppinterface/cb_summodel.cpp
dll void cb_summodel_destroy(SumModel* self) {
  delete self;
}

dll void cb_summodel_clear(SumModel* self) {
  self->clear();
}

dll SumModel* cb_summodel_new() {
  return new SumModel(0);
}

dll const SumBlockModel* cb_summodel_model(const SumModel* self, const FunctionObject* fo) {
  return self->model(fo);
}

dll Integer cb_summodel_nsubmodels(const SumModel* self) {
  return self->nsubmodels();
}

dll int cb_summodel_add_model(SumModel* self, SumBlockModel* model) {
  return self->add_model(model);
}

dll SumBlockModel* cb_summodel_remove_model(SumModel* self, const FunctionObject* fo) {
  return self->remove_model(fo);
}

dll int cb_summodel_eval_function(SumModel* self, Integer* ub_fid, Real* ub, Integer y_id, const Matrix* y, Real nullstep_bound, Real relprec) {
  return self->eval_function(*ub_fid, *ub, y_id, *y, nullstep_bound, relprec);
}

dll int cb_summodel_eval_model(SumModel* self, Real* lb, Integer y_id, const Matrix* y, Real relprec) {
  return self->eval_model(*lb, y_id, *y, relprec);
}

dll int cb_summodel_update_model(SumModel* self, int model_update, Integer center_id, const Matrix* center_y, Integer cand_id, const Matrix* cand_y, Real model_maxviol, BundleProxObject* H) {
  return self->update_model((ModelUpdate)model_update, center_id, *center_y, cand_id, *cand_y, model_maxviol, *H);
}

dll int cb_summodel_synchronize_ids(SumModel* self, Integer* new_center_ub_fid, Integer new_center_id, Integer old_center_id, Integer* new_cand_ub_fid, Integer new_cand_id, Integer old_cand_id, Integer* new_aggregate_id) {
  return self->synchronize_ids(*new_center_ub_fid, new_center_id, old_center_id, *new_cand_ub_fid, new_cand_id, old_cand_id, *new_aggregate_id);
}

dll int cb_summodel_center_modified(SumModel* self, Integer* function_modification_id, Integer center_id) {
  return self->center_modified(*function_modification_id, center_id);
}

dll int cb_summodel_recompute_center(SumModel* self, Integer* new_center_ub_fid, Real* new_center_ub, Integer center_id, const Matrix* y, int accept_only_higher_values = 0, Real relprec = -1.) {
  return self->recompute_center(*new_center_ub_fid, *new_center_ub, center_id, *y, (bool)accept_only_higher_values, relprec);
}

dll int cb_summodel_model_aggregate_modified(SumModel* self, Integer old_model_aggregate_id) {
  return self->model_aggregate_modified(old_model_aggregate_id);
}

dll int cb_summodel_provide_model_aggregate(SumModel* self, Integer y_id, const Matrix* y) {
  return self->provide_model_aggregate(y_id, *y);
}

dll int cb_summodel_add_variable_metric(SumModel* self, VariableMetric* H, Integer y_id, const Matrix* y, int descent_step, Real weightu, Real model_maxviol, const Indexmatrix* indices = 0) {
  return self->add_variable_metric(*H, y_id, *y, (bool)descent_step, weightu, model_maxviol, indices);
}

dll int cb_summodel_check_center_validity_by_candidate(SumModel* self, int* cand_minorant_is_below, Integer center_id, const Matrix* center_y) {
  cand_minorant_is_below = 0;
  return self->check_center_validity_by_candidate(*(bool*)cand_minorant_is_below, center_id, *center_y);
}

dll ModifiableOracleObject* cb_summodel_get_oracle_object(SumModel* self) {
  return self->get_oracle_object();
}

dll int cb_summodel_make_model_aggregate(SumModel* self, int* penalty_parameter_increased, int keep_penalty_fixed) {
  penalty_parameter_increased = 0;
  return self->make_model_aggregate(*(bool*)penalty_parameter_increased, (bool)keep_penalty_fixed);
}

dll Real cb_summodel_lb_function(SumModel* self, Integer y_id, const Matrix* y) {
  return self->lb_function(y_id, *y);
}

dll int cb_summodel_get_function_minorant(SumModel* self, MinorantPointer* minorant, const AffineFunctionTransformation* aft = 0) {
  return self->get_function_minorant(*minorant, aft);
}

dll int cb_summodel_get_center_minorant(SumModel* self, MinorantPointer* minorant, const AffineFunctionTransformation* aft = 0) {
  return self->get_center_minorant(*minorant, aft);
}

dll int cb_summodel_adjust_multiplier(SumModel* self, int* values_may_have_changed) {
  values_may_have_changed = 0;
  return self->adjust_multiplier(*(bool*)values_may_have_changed);
}

dll int cb_summodel_sumbundle_mode(SumModel* self, int* mode, SumBundleHandler* bh = 0, AffineFunctionTransformation* aft = 0) {
  return self->sumbundle_mode(*(SumBundle::Mode*)mode, bh, aft);
}

dll int cb_summodel_start_sumaugmodel(SumModel* self, QPModelDataPointer* blockp, Integer cand_id, const Matrix* cand_y, const Indexmatrix* indices = 0, SumBundleHandler* bh = 0, int mode = SumBundle::inactive, AffineFunctionTransformation* aft = 0) {
  return self->start_sumaugmodel(*blockp, cand_id, *cand_y, indices, bh, (SumBundle::Mode)mode, aft);
}

dll int cb_summodel_update_model2(SumModel* self, int model_update, Integer center_id, const Matrix* center_y, Integer cand_id, const Matrix* cand_y, Real null_step_factor, BundleProxObject* H, Real* model_deviation, Real* model_curvature) {
  return self->update_model((ModelUpdate)model_update, center_id, *center_y, cand_id, *cand_y, null_step_factor, *H, *model_deviation, *model_curvature);
}

dll BundleData* cb_summodel_get_data(SumModel* self) {
  return self->get_data();
}

dll const BundleData* cb_summodel_get_data2(const SumModel* self) {
  return self->get_data();
}

dll int cb_summodel_set_data(SumModel* self, BundleData* bd) {
  return self->set_data(bd);
}

dll const PrimalData* cb_summodel_get_approximate_primal(const SumModel* self) {
  return self->get_approximate_primal();
}

dll const PrimalData* cb_summodel_get_center_primal(const SumModel* self) {
  return self->get_center_primal();
}

dll const PrimalData* cb_summodel_get_candidate_primal(const SumModel* self) {
  return self->get_candidate_primal();
}

dll int cb_summodel_call_primal_extender(SumModel* self, PrimalExtender* param0) {
  return self->call_primal_extender(*param0);
}

dll int cb_summodel_set_bundle_parameters(SumModel* self, const BundleParameters* bp) {
  return self->set_bundle_parameters(*bp);
}

dll const BundleParameters* cb_summodel_get_bundle_parameters(const SumModel* self) {
  return self->get_bundle_parameters();
}

dll void cb_summodel_clear_model(SumModel* self, int discard_minorants_only = 0) {
  self->clear_model((bool)discard_minorants_only);
}

dll void cb_summodel_clear_aggregates(SumModel* self) {
  self->clear_aggregates();
}

dll void cb_summodel_set_out(SumModel* self, int pril = 1) {
  self->set_out(&std::cout, pril);
}

dll CH_Tools::Microseconds* cb_summodel_new_get_preeval_time(const SumModel* self) {
  return new CH_Tools::Microseconds(self->get_preeval_time());
}

dll CH_Tools::Microseconds* cb_summodel_new_get_eval_time(const SumModel* self) {
  return new CH_Tools::Microseconds(self->get_eval_time());
}

dll CH_Tools::Microseconds* cb_summodel_new_get_posteval_time(const SumModel* self) {
  return new CH_Tools::Microseconds(self->get_posteval_time());
}

